File: android/app/src/main/java/com/pms_parkin_mobile/util/Hex.java
package com.pms_parkin_mobile.util;

public class Hex {

    private static final char[] HEX_STRING_LOWER = {'0', '1', '2', '3', '4', '5', '6', '7', '8', '9','a', 'b', 'c', 'd', 'e', 'f'};
    private static final char[] HEX_STRING_UPPER = {'0', '1', '2', '3', '4', '5', '6', '7', '8', '9','A', 'B', 'C', 'D', 'E', 'F'};
    public static final int HEX_LOWER = 0;
    public static final int HEX_UPPER = 1;



    public static String bytesToHex(byte[] bytes, int start, int length, int hexType) {
        if (bytes == null || bytes.length <= start + length) {
            throw new IllegalArgumentException("bytes cannot be null or invalid length");
        }

        char[] HEX = HEX_STRING_LOWER;
        if (hexType == 1) {
            HEX = HEX_STRING_UPPER;
        }

        char[] result = new char[2 * length];
        int j = 0;
        for (int i = start; i < start + length; i++) {
            result[j++] = HEX[(0xf0 & bytes[i]) >>> 4];
            result[j++] = HEX[(0x0f & bytes[i])];
        }
        return new String(result);
    }
}

File: android/app/src/main/java/com/pms_parkin_mobile/util/Hex.kt
package com.pms_parkin_mobile.util

object Hex {
    private val HEX_STRING_LOWER =
        charArrayOf('0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'b', 'c', 'd', 'e', 'f')
    private val HEX_STRING_UPPER =
        charArrayOf('0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'A', 'B', 'C', 'D', 'E', 'F')
    const val HEX_LOWER: Int = 0
    const val HEX_UPPER: Int = 1


    fun bytesToHex(bytes: ByteArray, start: Int, length: Int, hexType: Int): String {
        require(!(bytes == null || bytes.size <= start + length)) { "bytes cannot be null or invalid length" }

        var HEX = HEX_STRING_LOWER
        if (hexType == 1) {
            HEX = HEX_STRING_UPPER
        }

        val result = CharArray(2 * length)
        var j = 0
        for (i in start..<start + length) {
            result[j++] = HEX[(0xf0 and bytes[i].toInt()) ushr 4]
            result[j++] = HEX[(0x0f and bytes[i].toInt())]
        }
        return String(result)
    }
}
